File: src/lede/_pins.py
"""v0.4.2 heading/pin retention helpers (pure, deterministic).

Locates headings by index in a sentence list and renders pinned content.
Reuses lede._headings for all "what is a heading" logic; adds no new
heading detection. The Rust mirror is rust/src/pins.rs.
"""
from __future__ import annotations

from collections.abc import Sequence

from lede._headings import is_structural_heading, md_depth


def nearest_heading_map(sentences: list[str]) -> list[int | None]:
    """For each sentence index, the index of the most recent heading
    sentence at or before it, or None if no heading precedes it.

    A heading's own entry is None (headings are never selected, so the
    value is unused; None keeps the rule "the heading above me")."""
    out: list[int | None] = []
    current: int | None = None
    for i, s in enumerate(sentences):
        if is_structural_heading(s):
            out.append(None)
            current = i
        else:
            out.append(current)
    return out


def document_title_index(sentences: list[str]) -> int | None:
    """Index of the document title: a depth-1 heading appearing before
    any body sentence. None if the first sentence-ish line is body, or
    the first heading is not markdown depth 1."""
    for i, s in enumerate(sentences):
        if is_structural_heading(s):
            return i if md_depth(s) == 1 else None
        return None
    return None


def render_toc_from_list(headings: Sequence[str]) -> str:
    """TOC from caller-supplied headings: deduped (first wins), given order,
    one per line, no indentation."""
    seen: set[str] = set()
    out: list[str] = []
    for h in headings:
        if h not in seen:
            seen.add(h)
            out.append(h)
    return "\n".join(out)


def render_toc(text: str) -> str:
    """Render a table of contents from the document outline. Each
    section name on its own line, indented (depth-1)*2 spaces."""
    from lede.extract.outline import outline

    lines: list[str] = []
    for section in outline(text):
        indent = "  " * max(0, section.depth - 1)
        lines.append(f"{indent}{section.name}")
    return "\n".join(lines)


def _render_keep_headings_override(
    sentences: list[str],
    selected: list[int],
    headings: Sequence[str],
) -> tuple[str, list[str]]:
    """Render body with caller-supplied headings per §4.3 of the spec.

    Returns (body_string, pinned_list) where pinned_list is every heading
    emitted in emission order."""
    seen: set[str] = set()
    H: list[str] = []
    for h in headings:
        if h not in seen:
            seen.add(h)
            H.append(h)

    matched_pos: dict[str, int] = {}
    claimed: set[int] = set()
    for h in H:
        hs = h.strip()
        for i, s in enumerate(sentences):
            if i in claimed:
                continue
            if s.strip() == hs:
                matched_pos[h] = i
                claimed.add(i)
                break
    pos_to_heading = {idx: h for h, idx in matched_pos.items()}
    matched_indices = sorted(matched_pos.values())

    pinned: list[str] = []
    out_lines: list[str] = []
    buf: list[str] = []
    emitted_idx: set[int] = set()
    emitted_text: set[str] = set()

    def flush() -> None:
        if buf:
            out_lines.append(" ".join(buf))
            buf.clear()

    title = H[0] if H else None
    if title is not None:
        out_lines.append(title)
        pinned.append(title)
        emitted_text.add(title)
        if title in matched_pos:
            emitted_idx.add(matched_pos[title])

    for h in H:
        if h in emitted_text:
            continue
        if h not in matched_pos:
            out_lines.append(h)
            pinned.append(h)
            emitted_text.add(h)

    def nearest_matched(s_idx: int) -> int | None:
        best: int | None = None
        for mi in matched_indices:
            if mi <= s_idx:
                best = mi
            else:
                break
        return best

    for s_idx in selected:
        mi = nearest_matched(s_idx)
        if mi is not None and mi not in emitted_idx:
            flush()
            h = pos_to_heading[mi]
            out_lines.append(h)
            pinned.append(h)
            emitted_idx.add(mi)
            emitted_text.add(h)
        buf.append(sentences[s_idx])
    flush()

    return "\n".join(out_lines), pinned


def render_with_pins(
    sentences: list[str],
    selected: list[int],
    *,
    keep_headings: bool,
    include_toc: bool,
    pin: Sequence[str] | None,
    text: str,
    headings: Sequence[str] | None = None,
) -> tuple[str, tuple[str, ...]]:
    """Weave pinned content around the extractive body.

    Returns (output_string, pinned_headings). `selected` must be sorted
    ascending. Headings interleave in document order; the title (if any)
    is pinned at the top; `pin` lines and the TOC prepend as blocks in
    the order pin -> toc -> body. Deterministic: dedup preserves first
    occurrence; no ordering depends on hashing or locale."""
    pinned_headings: list[str] = []

    if keep_headings:
        if headings:
            body, pinned_list = _render_keep_headings_override(sentences, selected, headings)
            pinned_headings = pinned_list
        else:
            heading_of = nearest_heading_map(sentences)
            title_idx = document_title_index(sentences)
            emitted: set[int] = set()
            out_lines: list[str] = []
            buf: list[str] = []

            def flush() -> None:
                if buf:
                    out_lines.append(" ".join(buf))
                    buf.clear()

            if title_idx is not None:
                out_lines.append(sentences[title_idx])
                emitted.add(title_idx)
                pinned_headings.append(sentences[title_idx])

            for s_idx in selected:
                h = heading_of[s_idx]
                if h is not None and h not in emitted:
                    flush()
                    out_lines.append(sentences[h])
                    emitted.add(h)
                    pinned_headings.append(sentences[h])
                buf.append(sentences[s_idx])
            flush()

            if pinned_headings:
                body = "\n".join(out_lines)
            else:
                # No headings pinned: byte-identical to the plain scorer.
                body = " ".join(sentences[i] for i in selected)
    else:
        body = " ".join(sentences[i] for i in selected)

    blocks: list[str] = []
    if pin:
        blocks.append("\n".join(pin))
    if include_toc:
        toc_text = render_toc_from_list(headings) if headings else render_toc(text)
        if toc_text:
            blocks.append(toc_text)
    blocks.append(body)
    output = "\n\n".join(b for b in blocks if b)
    return output, tuple(pinned_headings)


def prepend_blocks(
    body: str,
    *,
    include_toc: bool,
    pin: Sequence[str] | None,
    text: str,
) -> str:
    """Prepend pin and TOC blocks (in that order) to an already-rendered
    body string. Used when keep_headings is off but pin/include_toc are
    on, so the body stays byte-identical to the non-pinned scorer."""
    blocks: list[str] = []
    if pin:
        blocks.append("\n".join(pin))
    if include_toc:
        toc_text = render_toc(text)
        if toc_text:
            blocks.append(toc_text)
    if body:
        blocks.append(body)
    return "\n\n".join(b for b in blocks if b)

File: rust/src/pins.rs
//! v0.4.2 heading/pin retention. Mirror of src/lede/_pins.py. Byte-identical
//! output is enforced by `rust/tests/fixtures.rs::v0_4_2_pins_byte_identical`.

use crate::headings::{is_structural_heading, md_depth};

#[must_use]
pub fn nearest_heading_map(sentences: &[String]) -> Vec<Option<usize>> {
    let mut out = Vec::with_capacity(sentences.len());
    let mut current: Option<usize> = None;
    for (i, s) in sentences.iter().enumerate() {
        if is_structural_heading(s) {
            out.push(None);
            current = Some(i);
        } else {
            out.push(current);
        }
    }
    out
}

#[must_use]
pub fn document_title_index(sentences: &[String]) -> Option<usize> {
    // Look only at the first element: if it's a depth-1 structural heading,
    // return its index; any other case (body sentence or deeper heading) → None.
    if let Some((i, s)) = sentences.iter().enumerate().next() {
        if is_structural_heading(s) {
            return if md_depth(s) == 1 { Some(i) } else { None };
        }
    }
    None
}

#[must_use]
pub fn render_toc_from_list(headings: &[String]) -> String {
    let mut seen = std::collections::HashSet::new();
    let mut out: Vec<String> = Vec::new();
    for h in headings {
        if seen.insert(h.clone()) {
            out.push(h.clone());
        }
    }
    out.join("\n")
}

#[must_use]
pub fn render_toc(text: &str) -> String {
    let sections = crate::extract::outline::outline(text);
    sections
        .iter()
        .map(|sec| {
            let indent = "  ".repeat(sec.depth.saturating_sub(1));
            format!("{indent}{}", sec.name)
        })
        .collect::<Vec<_>>()
        .join("\n")
}

fn render_keep_headings_override(
    sentences: &[String],
    selected: &[usize],
    headings: &[String],
) -> (String, Vec<String>) {
    // Dedup headings preserving order (first occurrence wins).
    let mut seen = std::collections::HashSet::new();
    let mut h_list: Vec<String> = Vec::new();
    for h in headings {
        if seen.insert(h.clone()) {
            h_list.push(h.clone());
        }
    }

    // First-unclaimed stripped-equality match for each heading.
    let mut matched_pos: std::collections::HashMap<String, usize> =
        std::collections::HashMap::new();
    let mut claimed: std::collections::HashSet<usize> = std::collections::HashSet::new();
    for h in &h_list {
        let hs = h.trim();
        for (i, s) in sentences.iter().enumerate() {
            if claimed.contains(&i) {
                continue;
            }
            if s.trim() == hs {
                matched_pos.insert(h.clone(), i);
                claimed.insert(i);
                break;
            }
        }
    }
    let pos_to_heading: std::collections::HashMap<usize, String> = matched_pos
        .iter()
        .map(|(h, &idx)| (idx, h.clone()))
        .collect();
    let mut matched_indices: Vec<usize> = matched_pos.values().copied().collect();
    matched_indices.sort_unstable();

    let mut pinned: Vec<String> = Vec::new();
    let mut out_lines: Vec<String> = Vec::new();
    let mut buf: Vec<String> = Vec::new();
    let mut emitted_idx: std::collections::HashSet<usize> = std::collections::HashSet::new();
    let mut emitted_text: std::collections::HashSet<String> = std::collections::HashSet::new();

    // Title at top (first heading in h_list).
    if let Some(title) = h_list.first() {
        out_lines.push(title.clone());
        pinned.push(title.clone());
        emitted_text.insert(title.clone());
        if let Some(&ti) = matched_pos.get(title) {
            emitted_idx.insert(ti);
        }
    }

    // Unmatched headings (except title already emitted) go before body.
    for h in &h_list {
        if emitted_text.contains(h) {
            continue;
        }
        if !matched_pos.contains_key(h) {
            out_lines.push(h.clone());
            pinned.push(h.clone());
            emitted_text.insert(h.clone());
        }
    }

    // Walk selected sentences; interleave matched headings at nearest match.
    for &s_idx in selected {
        // nearest matched index <= s_idx
        let nearest = matched_indices
            .iter()
            .rev()
            .find(|&&mi| mi <= s_idx)
            .copied();
        if let Some(mi) = nearest {
            if !emitted_idx.contains(&mi) {
                if !buf.is_empty() {
                    out_lines.push(buf.join(" "));
                    buf.clear();
                }
                let h = pos_to_heading[&mi].clone();
                out_lines.push(h.clone());
                pinned.push(h.clone());
                emitted_idx.insert(mi);
                emitted_text.insert(h);
            }
        }
        buf.push(sentences[s_idx].clone());
    }
    if !buf.is_empty() {
        out_lines.push(buf.join(" "));
    }

    // Survival guarantee (§4.3 step 6): any supplied heading not yet emitted —
    // it matched a body line but that section had no selected sentence — is
    // appended so every heading in H is present in the output.
    for h in &h_list {
        if !emitted_text.contains(h) {
            out_lines.push(h.clone());
            pinned.push(h.clone());
            emitted_text.insert(h.clone());
        }
    }

    (out_lines.join("\n"), pinned)
}

#[allow(clippy::too_many_arguments)]
#[must_use]
pub fn render_with_pins(
    sentences: &[String],
    selected: &[usize],
    keep_headings: bool,
    include_toc: bool,
    pin: Option<&[String]>,
    text: &str,
    headings: Option<&[String]>,
) -> (String, Vec<String>) {
    let mut pinned_headings: Vec<String> = Vec::new();
    let body: String;
    if keep_headings {
        // Dispatch to override renderer when caller-supplied headings are present.
        let use_override = matches!(headings, Some(h) if !h.is_empty());
        if use_override {
            let h_list = headings.unwrap();
            let (b, pinned) = render_keep_headings_override(sentences, selected, h_list);
            body = b;
            pinned_headings = pinned;
        } else {
            // Auto-detection path — unchanged.
            let heading_of = nearest_heading_map(sentences);
            let title_idx = document_title_index(sentences);
            let mut emitted = std::collections::HashSet::new();
            let mut out_lines: Vec<String> = Vec::new();
            let mut buf: Vec<String> = Vec::new();
            if let Some(t) = title_idx {
                out_lines.push(sentences[t].clone());
                emitted.insert(t);
                pinned_headings.push(sentences[t].clone());
            }
            for &s_idx in selected {
                if let Some(h) = heading_of[s_idx] {
                    if !emitted.contains(&h) {
                        if !buf.is_empty() {
                            out_lines.push(buf.join(" "));
                            buf.clear();
                        }
                        out_lines.push(sentences[h].clone());
                        emitted.insert(h);
                        pinned_headings.push(sentences[h].clone());
                    }
                }
                buf.push(sentences[s_idx].clone());
            }
            if !buf.is_empty() {
                out_lines.push(buf.join(" "));
            }
            if pinned_headings.is_empty() {
                body = selected
                    .iter()
                    .map(|&i| sentences[i].clone())
                    .collect::<Vec<_>>()
                    .join(" ");
            } else {
                body = out_lines.join("\n");
            }
        }
    } else {
        body = selected
            .iter()
            .map(|&i| sentences[i].clone())
            .collect::<Vec<_>>()
            .join(" ");
    }

    let mut blocks: Vec<String> = Vec::new();
    if let Some(p) = pin {
        if !p.is_empty() {
            blocks.push(p.join("\n"));
        }
    }
    if include_toc {
        let toc_text = match headings {
            Some(h) if !h.is_empty() => render_toc_from_list(h),
            _ => render_toc(text),
        };
        if !toc_text.is_empty() {
            blocks.push(toc_text);
        }
    }
    if !body.is_empty() {
        blocks.push(body);
    }
    (blocks.join("\n\n"), pinned_headings)
}

#[must_use]
pub fn prepend_blocks(
    body: &str,
    include_toc: bool,
    pin: Option<&[String]>,
    text: &str,
    headings: Option<&[String]>,
) -> String {
    let mut blocks: Vec<String> = Vec::new();
    if let Some(p) = pin {
        if !p.is_empty() {
            blocks.push(p.join("\n"));
        }
    }
    if include_toc {
        let toc_text = match headings {
            Some(h) if !h.is_empty() => render_toc_from_list(h),
            _ => render_toc(text),
        };
        if !toc_text.is_empty() {
            blocks.push(toc_text);
        }
    }
    if !body.is_empty() {
        blocks.push(body.to_string());
    }
    blocks.join("\n\n")
}
